File: py/src/aegis/budget.py
"""with_resource_budget: the resource-budget primitive (Python sibling).

Wraps a blocking awaitable with a declared budget, optionally routes to
a fallback when the budget is exhausted, and emits a structured event
to a host-supplied observer.

Cross-language contract is locked in pact.yaml at the repo root. The TS
sibling uses AbortController + AbortSignal; we use asyncio.Event +
asyncio.wait_for. The fn receives the cancellation event; it is the
fn's responsibility to check `event.is_set()` (analog of
`signal.aborted`) or attach handlers.

To match the TS error-name string (so cross-lang outcomes are
byte-identical), we name the cancellation exception "AbortError" via
the `name` property — even though it's structurally a Python class.
"""

from __future__ import annotations

import asyncio
import time
from contextvars import ContextVar
from typing import Any, Awaitable, Callable, TypeVar

from .errors import AegisConfigError, BudgetExceededError, FallbackFailedError
from .types import (
    AegisEvent,
    AegisObserver,
    Tags,
    WrapArgs,
    resource_class_rank,
)

T = TypeVar("T")

# ContextVar mirrors AsyncLocalStorage. Tags propagate across `await`
# boundaries because asyncio sets up a fresh `Context` per task that
# inherits the parent's vars; nested `with_resource_budget` calls
# overwrite within their own scope and restore on exit (we use
# `set` + `reset(token)`).
_TAGS_CTX: ContextVar[Tags | None] = ContextVar("aegis_tags", default=None)


def get_aegis_tags() -> Tags | None:
    """Return active aegis tags from the current async context, or None."""
    return _TAGS_CTX.get()


# Cancellation exception that the wrapped fn raises when the budget
# fires. We name it `AbortError` so cross-language outcomes match the
# TS sibling's `err.name === 'AbortError'` check.
class AbortError(Exception):
    """Raised inside the wrapped fn when the budget timer fires."""

    @property
    def name(self) -> str:
        return "AbortError"

    def __init__(self) -> None:
        super().__init__("aborted")


# Host-installed observer. Default no-op so the primitive is usable in
# tests / sandboxes without wiring anything. Hosts (Reeve, Baton, etc.)
# call set_aegis_observer() at boot.
_observer: AegisObserver = lambda _evt: None  # noqa: E731


def set_aegis_observer(observer: AegisObserver) -> None:
    global _observer
    _observer = observer


def clear_aegis_observer() -> None:
    global _observer
    _observer = lambda _evt: None  # noqa: E731


def _emit(event: AegisEvent) -> None:
    """Dispatch to the host observer, swallowing its exceptions."""
    try:
        _observer(event)
    except Exception:
        # Observer failures are not load-bearing — instrumentation
        # outage cannot affect the wrapped fn's outcome.
        pass


def _make_abort_event() -> asyncio.Event:
    """Create the cancellation event the wrapped fn observes."""
    return asyncio.Event()


def _is_abort_like(exc: BaseException) -> bool:
    """Return True iff exc represents a cancellation/abort.

    Includes our internal AbortError, asyncio.CancelledError, and any
    exception whose class name is "AbortError" / "TimeoutError" (so the
    fn can raise its own analog without us needing to know its module).
    """
    if isinstance(exc, AbortError):
        return True
    if isinstance(exc, asyncio.CancelledError):
        return True
    cls_name = type(exc).__name__
    return cls_name in ("AbortError", "TimeoutError")


async def with_resource_budget(
    args: WrapArgs,
    fn: Callable[[asyncio.Event], Awaitable[T]],
) -> T:
    """Wrap a blocking awaitable with a declared budget.

    See pact.yaml `contract.outcomes` for the full state machine.
    """
    # Wrap-time validation: fallback resourceClass must not outrank primary's.
    if args.fallback is not None:
        primary_rank = resource_class_rank(args.primary_resource_class)
        fallback_rank = resource_class_rank(args.fallback.resource_class)
        if fallback_rank > primary_rank:
            raise AegisConfigError(
                f"fallback resourceClass={args.fallback.resource_class} "
                f"(rank {fallback_rank}) is more expensive than "
                f"primary={args.primary_resource_class} (rank {primary_rank}) "
                f"for {args.tags.component}.{args.tags.op}. "
                f"Fallback paths cannot escalate resource use."
            )

    abort_event = _make_abort_event()
    timeout_s = args.budget.timeout_ms / 1000.0
    t0 = time.monotonic()

    # Establish the tag context. We use a fresh token so on exit the
    # parent's tags (if any) are restored.
    tag_token = _TAGS_CTX.set(args.tags)

    # Schedule a timer that flips abort_event when the budget is up.
    loop = asyncio.get_running_loop()
    timer_handle: asyncio.Handle = loop.call_later(timeout_s, abort_event.set)

    try:
        try:
            # Run the primary fn under wait_for so we get a hard kill if
            # it ignores abort_event entirely. wait_for raises
            # asyncio.TimeoutError when the deadline is hit.
            result = await asyncio.wait_for(fn(abort_event), timeout=timeout_s)
            elapsed_ok = (time.monotonic() - t0) * 1000.0
            _emit(
                AegisEvent(
                    outcome="ok",
                    tags=args.tags,
                    elapsed_ms=elapsed_ok,
                    primary_resource_class=args.primary_resource_class,
                    timeout_ms=args.budget.timeout_ms,
                )
            )
            return result
        except BaseException as err:  # noqa: BLE001 — we re-raise after categorizing
            elapsed = (time.monotonic() - t0) * 1000.0
            is_abort = _is_abort_like(err)
            exhausted = abort_event.is_set() or elapsed >= args.budget.timeout_ms

            if is_abort and exhausted:
                # Budget exhausted. If a fallback is registered, run it
                # with a FRESH abort event (its own budget timer).
                if args.fallback is not None:
                    fb_event = _make_abort_event()
                    fb_timer = loop.call_later(timeout_s, fb_event.set)
                    try:
                        fb_result: Any = await asyncio.wait_for(
                            args.fallback.fn(fb_event), timeout=timeout_s
                        )
                        _emit(
                            AegisEvent(
                                outcome="fallback",
                                tags=args.tags,
                                elapsed_ms=elapsed,
                                primary_resource_class=args.primary_resource_class,
                                fallback_resource_class=args.fallback.resource_class,
                                timeout_ms=args.budget.timeout_ms,
                            )
                        )
                        return fb_result
                    except BaseException as fb_err:  # noqa: BLE001
                        _emit(
                            AegisEvent(
                                outcome="fallback_failed",
                                tags=args.tags,
                                elapsed_ms=elapsed,
                                primary_resource_class=args.primary_resource_class,
                                fallback_resource_class=args.fallback.resource_class,
                                error_message=str(err),
                                fallback_error_message=str(fb_err),
                                timeout_ms=args.budget.timeout_ms,
                            )
                        )
                        # Wrap both causes. Mirror the TS impl: the
                        # original cause is the primary's abort
                        # (AbortError); the fallback cause is whatever
                        # the fallback raised.
                        primary_cause = AbortError() if isinstance(err, asyncio.TimeoutError | asyncio.CancelledError) else err
                        fallback_cause: BaseException = fb_err
                        if isinstance(fb_err, asyncio.TimeoutError | asyncio.CancelledError):
                            fallback_cause = AbortError()
                        raise FallbackFailedError(
                            primary_cause, fallback_cause, args.tags
                        ) from fb_err
                    finally:
                        fb_timer.cancel()
                # No fallback — surface BudgetExceededError.
                _emit(
                    AegisEvent(
                        outcome="budget_exceeded",
                        tags=args.tags,
                        elapsed_ms=elapsed,
                        primary_resource_class=args.primary_resource_class,
                        timeout_ms=args.budget.timeout_ms,
                    )
                )
                raise BudgetExceededError(args.budget, elapsed, args.tags) from err

            # Primary threw something OTHER than abort, OR aborted
            # before budget. Don't swallow.
            _emit(
                AegisEvent(
                    outcome="primary_error",
                    tags=args.tags,
                    elapsed_ms=elapsed,
                    primary_resource_class=args.primary_resource_class,
                    error_message=str(err),
                    timeout_ms=args.budget.timeout_ms,
                )
            )
            raise
    finally:
        timer_handle.cancel()
        _TAGS_CTX.reset(tag_token)

File: src/budget.ts
// withResourceBudget: the resource-budget primitive. Wraps a blocking
// call with a declared budget, optionally routes to a fallback when the
// budget is exhausted, and emits a structured event to a host-supplied
// observer.
//
// Contract (see ../../pact.yaml for full text):
//   - Containable: cooperative async, fetch with AbortSignal, pg query
//     with statement_timeout. The primitive's timer fires, the wrapped
//     fn observes the signal, returns or rejects, the budget is honored.
//   - Uncontainable cooperatively: synchronous CPU burn. The primitive
//     sets the timer but the burn ignores it. Caller MUST set
//     `signalRespecting: false` when wrapping such code AND consider
//     worker_threads for actual containment.
//
// Multi-tenant addition: tags propagate via AsyncLocalStorage so nested
// calls inherit the tenant context.
//
// Observability: the host wires an AegisObserver via setAegisObserver().
// aegis does not import any logging library — that is consumer choice.

import { AsyncLocalStorage } from 'node:async_hooks';
import { AegisConfigError, BudgetExceededError, FallbackFailedError } from './errors.js';
import type { AegisEvent, AegisObserver, Tags, WrapArgs } from './types.js';
import { resourceClassRank } from './types.js';

const tagsContext = new AsyncLocalStorage<Tags>();

// Read the active aegis tags from the current async scope. Returns
// undefined if no aegis-wrapped fn is on the stack.
export function getAegisTags(): Tags | undefined {
  return tagsContext.getStore();
}

// Host-installed observer. Default is a no-op so the primitive remains
// usable in tests / sandboxes that don't wire one. Hosts (Reeve, Baton,
// etc.) call setAegisObserver() at boot to plug in their structured
// logger / tracer.
let observer: AegisObserver = () => {};

export function setAegisObserver(o: AegisObserver): void {
  observer = o;
}

// Reset to the no-op observer. Useful for tests.
export function clearAegisObserver(): void {
  observer = () => {};
}

function emit(event: AegisEvent): void {
  // Observer is best-effort. Swallow its exceptions — the wrapped fn's
  // result must not depend on whether instrumentation succeeded.
  try {
    observer(event);
  } catch {
    // intentional: observer failures are not load-bearing
  }
}

export async function withResourceBudget<T>(
  args: WrapArgs<T>,
  fn: (signal: AbortSignal) => Promise<T>,
): Promise<T> {
  // Wrap-time validation: fallback resourceClass must not outrank primary's.
  if (args.fallback) {
    const primaryRank = resourceClassRank(args.primaryResourceClass);
    const fallbackRank = resourceClassRank(args.fallback.resourceClass);
    if (fallbackRank > primaryRank) {
      throw new AegisConfigError(
        `fallback resourceClass=${args.fallback.resourceClass} (rank ${fallbackRank}) ` +
          `is more expensive than primary=${args.primaryResourceClass} (rank ${primaryRank}) ` +
          `for ${args.tags.component}.${args.tags.op}. Fallback paths cannot escalate resource use.`,
      );
    }
  }

  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(), args.budget.timeoutMs);
  const t0 = performance.now();

  try {
    return await tagsContext.run(args.tags, async () => {
      try {
        const result = await fn(ac.signal);
        const elapsedOk = performance.now() - t0;
        emit({
          outcome: 'ok',
          tags: args.tags,
          elapsedMs: elapsedOk,
          primaryResourceClass: args.primaryResourceClass,
          timeoutMs: args.budget.timeoutMs,
        });
        return result;
      } catch (err) {
        const elapsed = performance.now() - t0;
        const isAbort =
          err instanceof Error && (err.name === 'AbortError' || err.name === 'TimeoutError');
        const exhausted = ac.signal.aborted || elapsed >= args.budget.timeoutMs;

        if (isAbort && exhausted) {
          // Budget exhausted. If fallback is registered, invoke it with a
          // FRESH signal (so it gets its own budget rather than inheriting
          // the exhausted one).
          if (args.fallback) {
            const fbAc = new AbortController();
            const fbTimer = setTimeout(() => fbAc.abort(), args.budget.timeoutMs);
            try {
              const fbResult = await args.fallback.fn(fbAc.signal);
              emit({
                outcome: 'fallback',
                tags: args.tags,
                elapsedMs: elapsed,
                primaryResourceClass: args.primaryResourceClass,
                fallbackResourceClass: args.fallback.resourceClass,
                timeoutMs: args.budget.timeoutMs,
              });
              return fbResult;
            } catch (fbErr) {
              const fbErrMsg = fbErr instanceof Error ? fbErr.message : String(fbErr);
              emit({
                outcome: 'fallback_failed',
                tags: args.tags,
                elapsedMs: elapsed,
                primaryResourceClass: args.primaryResourceClass,
                fallbackResourceClass: args.fallback.resourceClass,
                errorMessage: (err as Error).message,
                fallbackErrorMessage: fbErrMsg,
                timeoutMs: args.budget.timeoutMs,
              });
              throw new FallbackFailedError(err as Error, fbErr as Error, args.tags);
            } finally {
              clearTimeout(fbTimer);
            }
          }
          // No fallback. Surface the budget-exceeded error.
          emit({
            outcome: 'budget_exceeded',
            tags: args.tags,
            elapsedMs: elapsed,
            primaryResourceClass: args.primaryResourceClass,
            timeoutMs: args.budget.timeoutMs,
          });
          throw new BudgetExceededError(args.budget, elapsed, args.tags);
        }

        // The primary fn threw something OTHER than abort, OR aborted
        // before the budget was up (fn cancelled itself for its own
        // reasons). Don't swallow — it's the caller's error to handle.
        emit({
          outcome: 'primary_error',
          tags: args.tags,
          elapsedMs: elapsed,
          primaryResourceClass: args.primaryResourceClass,
          errorMessage: err instanceof Error ? err.message : String(err),
          timeoutMs: args.budget.timeoutMs,
        });
        throw err;
      }
    });
  } finally {
    clearTimeout(timer);
  }
}
